cpp/src/groupby/sort/groupby.cu

```diff
@@ -0,0 +1,327 @@
+    update_column(final_value_columns[i]);
+  }
+  return final_value_columns;
+}
+
+template <bool keys_have_nulls, bool values_have_nulls>
+cudf::table process_original_requests(const cudf::table &input_keys,
+                               const Options &options,
+                               cudf::detail::groupby &groupby,
+                               const std::vector<AggRequestType> &original_requests,
+                               const std::vector<SimpleAggRequestCounter> &simple_requests,
+                               const std::vector<gdf_column *> &simple_values_columns,
+                               const std::vector<operators> &simple_operators,
+                               cudaStream_t &stream) {
+  index_vector group_indices = groupby.group_indices();
+  gdf_column key_sorted_order = groupby.key_sorted_order();
+  index_vector group_labels = groupby.group_labels();
+  gdf_size_type num_groups = (gdf_size_type)group_indices.size();
+  
+  cudf::table simple_values_table{simple_values_columns};
+
+  cudf::table simple_output_values{
+      num_groups, target_dtypes(column_dtypes(simple_values_table), simple_operators),
+      column_dtype_infos(simple_values_table), values_have_nulls, false, stream};
+
+  initialize_with_identity(simple_output_values, simple_operators, stream);
+
+  auto d_input_values = device_table::create(simple_values_table);
+  auto d_output_values = device_table::create(simple_output_values, stream);
+  rmm::device_vector<operators> d_ops(simple_operators);
+
+  auto row_bitmask = cudf::row_bitmask(input_keys, stream);
+
+  cudf::util::cuda::grid_config_1d grid_params{input_keys.num_rows(), 256};
+
+  cudf::groupby::sort::aggregate_all_rows<keys_have_nulls, values_have_nulls><<<
+      grid_params.num_blocks, grid_params.num_threads_per_block, 0, stream>>>(
+      input_keys.num_rows(), *d_input_values, *d_output_values, (gdf_index_type *)key_sorted_order.data,
+          group_labels.data().get(),
+          d_ops.data().get(), row_bitmask.data().get(), options.ignore_null_keys);
+
+  // compute_simple_request
+  return compute_original_requests(
+      original_requests, simple_requests, simple_output_values, stream);
+}
+
+template <bool keys_have_nulls, bool values_have_nulls>
+auto compute_sort_groupby(cudf::table const& input_keys, cudf::table const& input_values,
+                          std::vector<operators> const& input_ops,
+                          std::vector<operation_args*> const& input_ops_args,
+                          Options options,
+                          cudaStream_t stream) {
+  auto include_nulls = not options.ignore_null_keys;
+  auto groupby = cudf::detail::groupby(input_keys, include_nulls, options.null_sort_behavior, options.input_sorted);
+  index_vector group_indices = groupby.group_indices();
+
+  if (group_indices.size() == 0) {
+    cudf::table output_values(0, target_dtypes(column_dtypes(input_values), input_ops), column_dtype_infos(input_values));
```

Comment:
Agree -- it's weird to call that table constructor with 0 rows, since it doesn't explicitly handle the 0 case. However I would be OK with modifying the constructor to handle 0 rows appropriately.